File: archive/c/c-plus-plus/roman-numeral.cpp
#include <algorithm>
#include <iostream>
#include <ranges>
#include <string_view>

[[noreturn]] void usage() {
    std::cerr << "Usage: please provide a string of roman numerals\n";
    std::exit(1);
}

[[noreturn]] void error() {
    std::cerr << "Error: invalid string of roman numerals\n";
    std::exit(1);
}

constexpr int value(char c) {
    switch (c) {
        case 'I':
            return 1;
        case 'V':
            return 5;
        case 'X':
            return 10;
        case 'L':
            return 50;
        case 'C':
            return 100;
        case 'D':
            return 500;
        case 'M':
            return 1000;
        default:
            return -1;
    }
}

constexpr bool is_valid(char c) { return value(c) != -1; }

int main(int argc, char* argv[]) {
    if (argc < 2) usage();

    std::string_view s = argv[1];

    if (s.empty()) {
        std::cout << 0 << '\n';
        return 0;
    }

    if (!std::ranges::all_of(s, is_valid)) error();

    int result = 0;

    for (std::size_t i = 0; i < s.size(); ++i) {
        int curr = value(s[i]);

        if (i + 1 < s.size()) {
            int next = value(s[i + 1]);

            if (curr < next) {
                result += next - curr;
                ++i;
                continue;
            }
        }

        result += curr;
    }

    std::cout << result << '\n';
}

File: archive/t/typescript/roman-numeral.ts
// Roman Numeral Conversion

const romanNumeralValues = {
  "I": 1,
  "V": 5,
  "X": 10,
  "L": 50,
  "C": 100,
  "D": 500,
  "M": 1000
};

function romanNumeralConversion(str?: string) {

  // Check for blank input
  if (str === undefined) {
    return "Usage: please provide a string of roman numerals";
  }

  // Check for empty string
  if (str === "") {
    return 0;
  }

  // Get valid characters
  const validValues = Object.keys(romanNumeralValues)

  // Check for invalid characters
  for (let i = 0; i < str.length; i++) {
    const char = str.charAt(i)
    if (char === undefined || validValues.indexOf(char) === -1) {
      return "Error: invalid string of roman numerals";
    }
  }

  // Calculating result (adding/subtracting)
  // 
  // Note: next will be undefined for the last element, it'll
  // still work though since 'any number' < undefined will
  // be false ... hacky - but works!
  //
  // Note 2: This does not account for wrongly formatted
  // Roman Numerals like IIIIIIIIIX

  let answer = 0;
  for (let i = 0; i < str.length; i++) {
    const curr = romanNumeralValues[str.charAt(i)];
    const next = romanNumeralValues[str.charAt(i+1)];
    if (curr < next) {
      answer -= curr;
    } else {
      answer += curr;
    }
  }

  return answer;
}


// CLI use needs to have node.js installed
// Run like so:
//   node roman-numeral.js IIIIIIXXVXIV


// Process arguments from CLI
const args = process.argv.slice(2);

if (args.length < 1) {
  console.error("Usage: please provide a string of roman numerals");
} else {
  console.info(romanNumeralConversion(args[0]))
}

